java/arcs/core/storage/keys/JoinStorageKey.kt

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.core.storage.keys
+
+import arcs.core.storage.StorageKey
+import arcs.core.storage.StorageKeyParser
+import arcs.core.storage.StorageKeyUtils
+import arcs.core.storage.embed
+
+/** Protocol to be used when the StorageKey is composed of multiple StorageKeys. */
+const val COMPOSITE_PROTOCOL = "join"
+
+/** Implementation for a composite [StorageKey] for joining entities. */
+class JoinStorageKey(
+    private val storageKeys: List<StorageKey>
+) : StorageKey(COMPOSITE_PROTOCOL) {
+    override fun toKeyString(): String {
+        val builder = StringBuilder()
+        builder.append("${storageKeys.size}/")
+        storageKeys.forEach { builder.append("{${it.embed()}}") }
+
+        return builder.toString()
+    }
+
+    override fun childKeyWithComponent(component: String): StorageKey {
+        TODO("Not yet implemented for JoinStorageKey")
+    }
+
+    companion object {
+        init {
+            StorageKeyParser.addParser(COMPOSITE_PROTOCOL, ::fromString)
+        }
+
+        /** Register [JoinStorageKey] with the [StorageKeyParser]. */
+        fun registerParser() {
+            StorageKeyParser.addParser(COMPOSITE_PROTOCOL, ::fromString)
+        }
+
+        private fun fromString(rawValue: String): JoinStorageKey {
+            val invalidFormatMessage: () -> String =
+                { "Invalid format for JoinStorageKey: $rawValue" }
+
+            // We will support < 10 joins.
+            val numberOfJoins: Int = rawValue[0] - '0'
```

Comment:
Given we can rely on the number of joined components based from the list alone, should we remove the 'N/' component entirely? I would support that. So we would instead have simply:

join://{key1}{key2}{key3}

what do you think?

Reply:
Maybe it's more readable and more explicit with `join://3/{key1}{key2}{key3}` ? I'm on the fence myself...

Reply:
We can handle that in the next cl.